python/srf/_pysrf/src/pipeline.cpp

```diff
@@ -18,42 +18,147 @@
 #include "pysrf/pipeline.hpp"
 
 #include "pysrf/types.hpp"
-#include "srf/pipeline/pipeline.hpp"
-#include "srf/segment/builder.hpp"
+#include "pysrf/utils.hpp"
 
+#include "srf/node/forward.hpp"
+#include "srf/node/port_registry.hpp"
+#include "srf/segment/builder.hpp"  // IWYU pragma: keep
+#include "srf/segment/egress_ports.hpp"
+#include "srf/segment/ingress_ports.hpp"
+#include "srf/segment/ports.hpp"
+
+#include <glog/logging.h>
+#include <pybind11/cast.h>
 #include <pybind11/gil.h>
-#include <pybind11/pybind11.h>
 #include <pybind11/pytypes.h>
+#include <tupleobject.h>
 
 #include <functional>
 #include <memory>
+#include <ostream>
+#include <stdexcept>
 #include <string>
+#include <tuple>
+#include <typeindex>
+#include <typeinfo>
 #include <utility>  // for move
+#include <vector>
 
-namespace {
-namespace py = pybind11;
-using namespace srf::pysrf;
+// IWYU pragma: no_include <pybind11/detail/common.h>
 
-template <std::size_t Count, template <class...> class PortClass, typename... ArgsT>
-struct Pipeline__PortBuilder : Pipeline__PortBuilder<Count - 1, PortClass, PyHolder, ArgsT...>
-{};
+namespace srf::pysrf {
+namespace py = pybind11;
 
-template <template <class...> class PortClass, typename... ArgsT>
-struct Pipeline__PortBuilder<0, PortClass, ArgsT...>
-{
-    using port_type_t                  = PortClass<ArgsT...>;
-    static constexpr size_t port_count = sizeof...(ArgsT);
+using ingress_info_t = std::tuple<std::vector<std::string>,
+                                  std::vector<std::type_index>,
+                                  std::vector<srf::node::PortUtil::ingress_builder_fn_t>>;
 
-    static port_type_t build(const std::vector<std::string>& port_ids)
-    {
-        return port_type_t(port_ids);
-    }
-};
-}  // namespace
+using egress_info_t = std::tuple<std::vector<std::string>,
+                                 std::vector<std::type_index>,
+                                 std::vector<srf::node::PortUtil::egress_builder_fn_t>>;
```

Comment:
Can we just make these their own struct so its easier to know what each field is.